### DISCUSSION TITLE
How to run server with callback and asnyc

### DISCUSSION THREAD

**Discussion by utack** [pymodbus-dev/pymodbus#2444] [Help] [answered]
Hi,

I am trying to implement a server with callback and an async concept.
Values provided over modbus are retrieved from a CAN device with async functions and should not block the rest of the application (uvicorn etc.)

However I have no idea how to get setValues and getValues to fit in.
I tried to use "async_setValues" and "async_getValues" instead of "setValues" and "getValues" but they are not called.

Is there a way I could make the code below work?

Thank you very much
Jakob

```
import asyncio
from pymodbus.datastore import ModbusSequentialDataBlock
from pymodbus.datastore import ModbusSlaveContext, ModbusServerContext
from pymodbus.server.async_io import StartAsyncSerialServer


class CallbackDataBlock(ModbusSequentialDataBlock):

    def async_setValues(self, address, values):
        await something(values)

    def async_getValues(self, address, count=1):
        return await something(count)


store = ModbusSlaveContext(
    di=CallbackDataBlock(0, [0] * 10000),
    co=CallbackDataBlock(0, [0] * 10000),
    hr=CallbackDataBlock(0, [0] * 10000),
    ir=CallbackDataBlock(0, [0] * 10000))
context = ModbusServerContext(slaves=store)

asyncio.run(StartAsyncSerialServer(context=context, port="COM17"))
```

**Comment by janiversen** [comment]
Have a look at the callback example, the app need to call set/get values not the other way round.

**Comment by janiversen** [comment]
Have a look at the callback example, the app need to call set/get values not the other way round.

**Comment by utack** [comment]
My bad,

it seems like I have had a misunderstanding about this example.
Is there an example that allows complete custom handling of all write and read requests?

Thank you again!

**Reply by janiversen** [reply]
No. If you want to catch all messages you need to look at the simulator, which allows to add trace functions.

Callback works that the app read/write data from/to the server

**Comment by ilkka-ollakka** [comment, marked as answer]
You could get forward with following changes:
- Instead of ModbusSequentialDataBlock use ModbusBaseSlaveContext as base-class of your own implementation
- reimplement async_getValues and async_setValues as you have done (check that you have defined them async, I assume it is typo in your example not to be there)
- create your class as slave-object directly instead of passing it to ModbusSlaveContext, so something like `context = ModbusServerContext(slaves=CallbackDataBlock())` in your example-code

I assume you can already connect to server via serial connection and all that is working as expected?

**Reply by janiversen** [reply]
It is a solution, that works, but it seems much easier to use the trace callbacks provided in simulator.
